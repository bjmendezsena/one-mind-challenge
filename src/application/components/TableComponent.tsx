import React from "react";
import { Table } from "antd";

import { useTable } from "../hooks/useTable";

export const TableComponent = () => {
  const { data, columns } = useTable();
  return (
    <div
      style={{
        margin: "10px 50px",
      }}
    >
      <Table
        columns={columns}
        dataSource={data}
        pagination={{
          defaultPageSize: 10,
          showSizeChanger: true,
          pageSizeOptions: ["10", "25"],
          hideOnSinglePage: true,
        }}
      />
    </div>
  );
};
