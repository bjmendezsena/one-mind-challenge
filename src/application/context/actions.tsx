import { UserInterface } from "../appInterfaces/appInterfaces";
export type Action =
  | {
      type: "START_GET_USERS";
    }
  | {
      type: "GET_USERS_FAILED";
      payload: string | null;
    }
  | {
      type: "GET_USERS_SUCCEED";
      payload: UserInterface[];
    }
  | {
      type: "ADD_USER";
      payload: UserInterface;
    }
  | {
      type: "REMOVE_USER";
      payload: UserInterface;
    }
  | {
      type: "START_GET_ANIMALS";
    }
  | {
      type: "GET_ANIMALS_FAILED";
      payload: string | null;
    }
  | {
      type: "GET_ANIMALS_SUCCEED";
      payload: string[];
    };
