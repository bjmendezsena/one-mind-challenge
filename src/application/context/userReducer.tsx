import { AppState } from "../appInterfaces/appInterfaces";
import { Action } from "./actions";

export const appReducer = (state: AppState, action: Action): AppState => {
  switch (action.type) {
    case "START_GET_USERS":
      return {
        ...state,
        isLoading: true,
      };

    case "GET_USERS_FAILED":
      return {
        ...state,
        isLoading: false,
        error: action.payload,
      };

    case "GET_USERS_SUCCEED":
      return {
        ...state,
        isLoading: false,
        users: action.payload,
        error: null,
      };

    case "ADD_USER":
      return {
        ...state,
        users: [...state.users, action.payload],
      };
    case "REMOVE_USER":
      return {
        ...state,
        users: state.users.filter((user) => user.id !== action.payload.id),
      };

    default:
      return state;
  }
};
