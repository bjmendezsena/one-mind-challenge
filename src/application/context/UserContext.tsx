import React, { createContext, useReducer, useState, useEffect } from "react";
import {
  UserInterface,
  AppState,
  UserDTO,
} from "../appInterfaces/appInterfaces";
import { fetchAllAnimals, fetchAllUSers } from "../services/dataService";
import { appReducer } from "./userReducer";

type ContextProps = {
  users: UserInterface[];
  isLoading: boolean;
  error: string | null;
  animals: string[];
  selectedAnimal: string;
  addUser: (userData: UserDTO) => void;
  removeUser: (user: UserInterface) => void;
  getAllUsers: () => void;
  getAllAnimals: () => void;
  setSelectedAnimal: (animal: string) => void;
};

export const AppContext = createContext({} as ContextProps);

const initialState: AppState = {
  error: null,
  isLoading: false,
  users: [],
  animals: [],
};

interface IAppProvider {
  children: JSX.Element | JSX.Element[];
}

export const AppProvider = ({ children }: IAppProvider) => {
  const [state, dispatch] = useReducer(appReducer, initialState);

  const [selectedAnimal, setSelectedAnimal] = useState<string>("");

  useEffect(() => {
    getAllUsers();
    getAllAnimals();
  }, []);
  const { animals } = state;
  useEffect(() => {
    if (animals.length) {
      setSelectedAnimal(animals[0]);
    }
  }, [animals]);

  const addUser = (userData: UserDTO) => {
    const { users } = state;
    const user: UserInterface = {
      ...userData,
      id: `${userData.given}${userData.surname}${users.length}`,
      name: {
        given: userData.given,
        surname: userData.surname,
      },
    };
    dispatch({
      type: "ADD_USER",
      payload: user,
    });
  };

  const removeUser = (user: UserInterface) => {
    dispatch({
      type: "REMOVE_USER",
      payload: user,
    });
  };

  const getAllAnimals = async () => {
    dispatch({
      type: "START_GET_ANIMALS",
    });
    const [animals, error] = await fetchAllAnimals();
    if (error) {
      return dispatch({
        type: "GET_ANIMALS_FAILED",
        payload:
          "An unexpected error occurred while trying to display the animals",
      });
    }
    dispatch({
      type: "GET_ANIMALS_SUCCEED",
      payload: animals,
    });
  };
  const getAllUsers = async () => {
    dispatch({
      type: "START_GET_USERS",
    });

    const [users, error] = await fetchAllUSers();
    if (error) {
      return dispatch({
        type: "GET_USERS_FAILED",
        payload:
          "An unexpected error occurred while trying to display the users",
      });
    }

    dispatch({
      type: "GET_USERS_SUCCEED",
      payload: users,
    });
  };

  return (
    <AppContext.Provider
      value={{
        ...state,
        addUser,
        removeUser,
        getAllUsers,
        getAllAnimals,
        selectedAnimal,
        setSelectedAnimal,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
