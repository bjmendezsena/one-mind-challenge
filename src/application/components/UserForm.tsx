import { Form, Input, Select, InputNumber, Switch } from "antd";
import React, { useContext } from "react";
import { AppContext } from "../context/UserContext";

import { CloseOutlined, CheckOutlined } from "@ant-design/icons";
import { UserDTO } from "../appInterfaces/appInterfaces";

const { Option } = Select;

interface IUserFormProps {
  onChange: (value: any, field: keyof UserDTO) => void;
  values: UserDTO;
}

export const UserForm = ({ onChange, values }: IUserFormProps) => {
  const { animals } = useContext(AppContext);

  return (
    <Form labelCol={{ span: 4 }} wrapperCol={{ span: 14 }} layout="horizontal">
      <Form.Item label="Name">
        <Input.Group compact>
          <Form.Item
            name={["name", "given"]}
            noStyle
            rules={[{ required: true, message: "Given name is required" }]}
          >
            <Input
              style={{ width: "50%" }}
              placeholder="Given"
              value={values.given}
              onChange={({ target }) => onChange(target.value, "given")}
            />
          </Form.Item>
          <Form.Item
            name={["name", "surname"]}
            noStyle
            rules={[{ required: true, message: "Surname is required" }]}
          >
            <Input
              style={{ width: "50%" }}
              placeholder="Surname"
              value={values.surname}
              onChange={({ target }) => onChange(target.value, "surname")}
            />
          </Form.Item>
        </Input.Group>
      </Form.Item>

      <Form.Item
        label="Points"
        rules={[{ required: true, message: "Points is required" }]}
      >
        <InputNumber
          style={{ width: "50%" }}
          value={values.points}
          onChange={(value) => onChange(value, "points")}
        />
      </Form.Item>
      <Form.Item
        label="Age"
        rules={[{ required: true, message: "Age is required" }]}
      >
        <InputNumber
          style={{ width: "50%" }}
          value={values.age}
          onChange={(value) => onChange(value, "age")}
        />
      </Form.Item>
      <Form.Item
        label="Animals"
        rules={[{ required: true, message: "Animals is required" }]}
      >
        <Select
          style={{ width: "100%" }}
          placeholder="Select animals"
          mode="multiple"
          defaultValue={values.animals}
          onChange={(values) => onChange(values, "animals")}
        >
          {React.Children.toArray(
            animals.map((animal) => <Option value={animal}>{animal}</Option>)
          )}
        </Select>
      </Form.Item>
      <Form.Item label="Active">
        <Switch
          checkedChildren={<CheckOutlined />}
          unCheckedChildren={<CloseOutlined />}
          checked={values.isActive}
          onChange={() => onChange(!values.isActive, "isActive")}
        />
      </Form.Item>
    </Form>
  );
};
