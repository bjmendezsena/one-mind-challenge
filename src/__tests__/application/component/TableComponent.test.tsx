import { shallow } from "enzyme";
import { TableComponent } from "../../../application/components/TableComponent";
import { AppProvider } from "../../../application/context/UserContext";

describe("Header tests", () => {
  it("Should render the component", async () => {
    const wrapper = shallow(
      <AppProvider>
        <TableComponent />
      </AppProvider>
    );

    expect(wrapper).toMatchSnapshot();
  });
});
