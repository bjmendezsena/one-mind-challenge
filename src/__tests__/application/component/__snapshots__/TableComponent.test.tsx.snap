// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header tests Should render the component 1`] = `
<ContextProvider
  value={
    Object {
      "addUser": [Function],
      "animals": Array [],
      "error": null,
      "getAllAnimals": [Function],
      "getAllUsers": [Function],
      "isLoading": false,
      "removeUser": [Function],
      "selectedAnimal": "",
      "setSelectedAnimal": [Function],
      "users": Array [],
    }
  }
>
  <TableComponent />
</ContextProvider>
`;
